examples/small-coalton-programs/src/brainfold.lisp

```diff
@@ -0,0 +1,341 @@
+              bf-array)
+    (current-bf-cell-value))
+
+  ;;
+  ;; Printing Cells (.)
+  ;;
+  
+  (declare print (Ufix -> String))
+  (define (print ascii-value)
+    "Converts a Ufix into its ascii value."
+    (into (make-list (unwrap (char:code-char ascii-value)))))
+
+  
+  (define (print-current-cell)
+    "Prints the current bf-cell."
+    (cell:write! print-buffer
+                 (str:concat (cell:read print-buffer)
+                             (print
+                              (unwrap
+                               (tryinto
+                                (current-bf-cell-value))))))
+    (current-bf-cell-value))
+
+  ;;
+  ;; Taking Input (,)
+  ;; 
+
+  ;; Currently takes individual characters one at a time as prompted
+
+  (define (character-prompt)
+    "A prompt for obtaining one character as input."
+    (Lisp Char ()
+      (cl:format cl:*query-io* "Input a character: ")
+      (cl:force-output cl:*query-io*)
+      (cl:coerce (cl:read-line cl:*query-io*)
+                 'cl:character)))
+  
+  (define (take-input)
+    "Takes and stores a character at the input-index as an ascii code at
+the pointer."
+    (vec:set! (cell:read pointer)
+              (into (char:char-code (character-prompt)))
+              bf-array)
+    (current-bf-cell-value)))
+
+;;;
+;;; Parsing/Lexing
+;;;
+
+(coalton-toplevel
+
+  (define-type Cmd
+    BFRight
+    BFLeft
+    BFPlus
+    BFMinus
+    BFPrint
+    BFInput
+    (BFLoop (List Cmd)))
+
+  (declare split-string (String -> List String))
+  (define (split-string s)
+    "Takes a string, returns a list of characters."
+    (if (arith:zero? (str:length s))
+        nil
+        (cons (substring s 0 1)
+              (split-string (substring s 1 (str:length s))))))
+
+  (declare index-after-loop ((List String) -> UFix))
+  (define (index-after-loop cmd-strings)
+    "Takes a list of bf strings, returns the count after the end of the
+corresponding loop."
+    (let ((until-end-loop-count
+            (fn (cmd-strings depth cmd-count)
+              (match cmd-strings
+                ((Nil) (error "Incomplete Loop"))
+                ((Cons "]" _)
+                 (if (arith:zero? depth)
+                     cmd-count
+                     (until-end-loop-count (list:cdr cmd-strings)
+                                           (1- depth)
+                                           (1+ cmd-count))))
+                ((Cons "[" _)
+                 (until-end-loop-count (list:cdr cmd-strings)
+                                       (1+ depth)
+                                       (1+ cmd-count)))
+                (_ (until-end-loop-count (list:cdr cmd-strings)
+                                         depth
+                                         (1+ cmd-count)))))))
+      (until-end-loop-count cmd-strings 0 2)))
+
+  (declare parse (String -> (List Cmd)))
+  (define (parse input-string)
+    "Parses Brainfold input into a lexical tree."
+    (let ((match-cmds
```

Comment:
Again, you might want to make this tail recursive. If it were me, I'd skip the string splitting altogether, and just traverse the string character by character, recursing when we hit a `[`. Then we don't need `split-string` or `index-after-loop`.